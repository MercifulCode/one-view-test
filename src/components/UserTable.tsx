import * as React from "react";
import { connect } from "react-redux";
import { TableRow } from "./TableRow";
import { fetchUserData } from "../reducers/JSONReducer";
import { bindActionCreators, Dispatch } from "redux";

export interface IUserTableProps {
  data: IUserResponse[];
  dispatchJSONFetch(fetchFn: any): void;
  onClick(currentUser: IUserResponse): void;
}

export interface IUserTableState {
  nameFilter: string;
}

export class UserTableClass extends React.Component<
  IUserTableProps,
  IUserTableState
> {
  constructor(props: IUserTableProps) {
    super(props);
    this.state = {
      nameFilter: "",
    };
  }

  public componentDidMount() {
    this.props.dispatchJSONFetch("https://jsonplaceholder.typicode.com/users");
  }

  public render() {
    return (
      <div>
        {this.renderSearchInput()}
        {this.renderTable()}
      </div>
    );
  }

  protected renderTable() {
    const { data, onClick } = this.props;
    const { nameFilter } = this.state;

    return (
      <table>
        <thead>
          <TableRow cellContent={["Name", "Email", "City", "Company"]} />
        </thead>
        <tbody>
          {data.map(
            (datum, index) =>
              datum.name
                .toLocaleLowerCase()
                .includes(nameFilter.toLocaleLowerCase()) && (
                <TableRow
                  key={`user_table_row_${index}`}
                  cellContent={[
                    datum.name,
                    datum.email,
                    datum.address.city,
                    datum.company.name,
                  ]}
                  onClick={() => {
                    onClick(datum);
                  }}
                />
              )
          )}
        </tbody>
      </table>
    );
  }

  protected renderSearchInput() {
    const { nameFilter } = this.state;
    return (
      <form onSubmit={this.handleSubmit}>
        <label>
          Filter By Name:
          <br />
          <input type="text" value={nameFilter} onChange={this.handleChange} />
        </label>
        <input type="submit" value="Submit" />
      </form>
    );
  }

  protected handleSubmit = (event: React.ChangeEvent<HTMLFormElement>) => {
    this.setState({ nameFilter: event.target.value });
  };

  protected handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    this.setState({ nameFilter: event.target.value });
  };
}

const mapStateToProps = (state: { userData: IUserResponse[] }) => ({
  data: state.userData,
});

const mapDispatchToProps = (dispatch: Dispatch) =>
  bindActionCreators(
    {
      dispatchJSONFetch: fetchUserData,
    },
    dispatch
  );

export const UserTable = connect(
  mapStateToProps,
  mapDispatchToProps
)(UserTableClass);
