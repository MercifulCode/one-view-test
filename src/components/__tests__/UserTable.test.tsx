import { UserTableClass } from "../UserTable";
import { shallow } from "enzyme";

describe("UserTable", () => {
  const aUser: IUserResponse = {
    address: {
      city: "Springfield",
      street: "Fake St",
      suite: "123",
      zipcode: "12345",
      geo: {
        lat: "0",
        lng: "0",
      },
    },
    name: "Homer Simpson",
    company: {
      name: "Nuclear Power Plant",
      catchPhrase: "Excellent",
      bs: "Mmmmmm bees",
    },
    email: "chunkylover53@aol.com",
    id: 1,
    phone: "123-456-7890",
    username: "HomerJay",
    website: "",
  };
  it("Should render when given no data", () => {
    const wrapper = shallow(
      <UserTableClass
        data={[]}
        dispatchJSONFetch={jest.fn()}
        onClick={jest.fn()}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("Should render when given some data", () => {
    const wrapper = shallow(
      <UserTableClass
        data={[aUser]}
        dispatchJSONFetch={jest.fn()}
        onClick={jest.fn()}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
});
