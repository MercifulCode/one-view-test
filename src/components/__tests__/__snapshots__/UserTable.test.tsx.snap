// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserTable Should render when given no data 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <label>
      Filter By Name:
      <br />
      <input
        onChange={[Function]}
        type="text"
        value=""
      />
    </label>
    <input
      type="submit"
      value="Submit"
    />
  </form>
  <table>
    <thead>
      <TableRow
        cellContent={
          Array [
            "Name",
            "Email",
            "City",
            "Company",
          ]
        }
      />
    </thead>
    <tbody />
  </table>
</div>
`;

exports[`UserTable Should render when given some data 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <label>
      Filter By Name:
      <br />
      <input
        onChange={[Function]}
        type="text"
        value=""
      />
    </label>
    <input
      type="submit"
      value="Submit"
    />
  </form>
  <table>
    <thead>
      <TableRow
        cellContent={
          Array [
            "Name",
            "Email",
            "City",
            "Company",
          ]
        }
      />
    </thead>
    <tbody>
      <TableRow
        cellContent={
          Array [
            "Homer Simpson",
            "chunkylover53@aol.com",
            "Springfield",
            "Nuclear Power Plant",
          ]
        }
        key="user_table_row_0"
        onClick={[Function]}
      />
    </tbody>
  </table>
</div>
`;
